core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp

```diff
@@ -57,6 +57,80 @@ namespace Kokkos {
 namespace Experimental {
 namespace Impl {
 
+enum class BlockType { Max, Preferred };
+
+template <typename DriverType, typename LaunchBounds = Kokkos::LaunchBounds<>,
+          HIPLaunchMechanism LaunchMechanism =
+              DeduceHIPLaunchMechanism<DriverType>::launch_mechanism>
+unsigned get_preferred_blocksize_impl() {
+  // FIXME_HIP - could be if constexpr for c++17
+  if (!HIPParallelLaunch<DriverType, LaunchBounds,
+                         LaunchMechanism>::default_launchbounds()) {
+    // use the user specified value
+    return LaunchBounds::maxTperB;
+  } else {
+    if (HIPParallelLaunch<DriverType, LaunchBounds,
+                          LaunchMechanism>::get_scratch_size() > 0) {
+      return HIPTraits::ConservativeThreadsPerBlock;
+    }
+    return HIPTraits::MaxThreadsPerBlock;
```

Comment:
uhm doesn't that mean we will spill like crazy?

Reply:
All of this is just code that has been moved around from https://github.com/kokkos/kokkos/pull/3953 We automatically adapt the blocksize to decrease the spilling.